### PR TITLE
Allow usage of ResizeObserver in ResizeSensor

### PR DESCRIPTION
**Pull request by jacehensley-wf**
## Ultimate problem:
Different approach to https://github.com/Workiva/over_react/pull/130 using this [ResizeObserver](https://github.com/JaceHensley/resize_observer_polyfill_wrapper) instead.

This approach will work in all browsers, but it doesn't use the built in functionality in Chrome.

## How it was fixed:
- Add a `props.useResizeObserver` that defaults to `false`
- Refactor `ResizeSensor` to render one of two internal impls
- Update tests

## Testing suggestions:
- Verify tests pass
- Pull this down and mess around with the new `ResizeObserver` component and verify that it works as expected.

## Potential areas of regression:
- `ResizeSensor`

---
__FYA:__ @greglittlefield-wf @aaronlademann-wf @jacehensley-wf @clairesarsam-wf @joelleibow-wf
__FYI:__ @robbecker-wf

### CODE REVIEW

**Comment by aviary2-wf** [comment]
## Raven
<details>
<summary>Number of Findings: 0</summary>

</details>

File: lib/src/component/resize/resize_props_mixin.dart [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  ///
+  ///     * [onInitialize] will never be called.
+  ///
+  /// * The sensors will be initialized/reset in the next animation frame after mount, as opposed to synchronously,
+  ///   helping to break up resulting layouts.
+  ///
+  /// Default: false
+  bool quickMount;
+
+  /// Whether to use a polyfilled `ResizeObserver`.
```
Comment (aaronlademann-wf) [inline lib/src/component/resize/resize_props_mixin.dart]: Probably should put a bit more meat on this bone to explain the benefits, drawbacks, etc.

File: lib/src/component/resize/resize_sensor.dart [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+/// A wrapper component that detects when its parent is resized.
+///
+/// This component _must_ be put in a relative or absolutely positioned
+/// container.
+///
+///     (ResizeSensor()..onResize = () => print('resized'))(children)
+///
+/// See: <https://docs.workiva.org/web_skin_dart/latest/components/#resize-sensor>.
+@Factory()
+UiFactory<ResizeSensorProps> ResizeSensor;
```
Comment (aaronlademann-wf) [inline lib/src/component/resize/resize_sensor.dart]: IMO its a little confusing that this factory / props class / component class are named identically to the one we export.

File: pubspec.yaml [hunk 3]
```diff
@@ -40,5 +42,6 @@ transformers:
       ]
   - test/pub_serve:
       $include: test/**_test{.*,}.dart
+  - dart_to_js_script_rewriter
```
Comment (aaronlademann-wf) [inline pubspec.yaml]: ?

File: test/over_react_component_test.html [hunk 4]
```diff
@@ -21,6 +21,7 @@
     <!-- javascript -->
     <script src="packages/react/react_with_addons.js"></script>
     <script src="packages/react/react_dom.js"></script>
+    <script src="packages/resize_observer_polyfill_wrapper/resize_observer_polyfill.js"></script>
```
Comment (aaronlademann-wf) [inline test/over_react_component_test.html]: Is this JS file necessary if `props.useResizeObserver` is `true`?

Is there any sort of validation within the component enforcing this?

If not - at the very least - there should be extensive documentation in the doc comment for `props.useResizeObserver` explaining that this is a requirement, IMO.

File: lib/src/component/resize/resize_sensor.dart (hunk 2 above)
Comment (jacehensley-wf) [reply lib/src/component/resize/resize_sensor.dart]: I agree, do you have a suggestion?

File: test/over_react_component_test.html (hunk 4 above)
Comment (jacehensley-wf) [reply test/over_react_component_test.html]: It's only necessary if that is `true`.  There's no validation here, just my underlying [package](https://github.com/JaceHensley/resize_observer_polyfill_wrapper)

Comment (robbecker-wf) [reply test/over_react_component_test.html]: So, would we need to include this JS file in the main shell .html file? Hopefully with feature detection and only load it if it needs the polyfill?

Comment (jacehensley-wf) [reply test/over_react_component_test.html]: The JS library that I wrapped never uses the native ResizeObserver, for whatever reason, so this JS file will always need to be loaded.

Comment (robbecker-wf) [reply test/over_react_component_test.html]: oh, that's weird. Isn't the whole point of using the browser built in resize observer to reduce code? This change just replaces the resize detection logic from our Dart polyfill to a JS implemented polyfill then?

Comment (jacehensley-wf) [reply test/over_react_component_test.html]: Kinda, but it's a lot cleaner of an impl IMO. It doesn't wrap the resizing element and place "resize sensors" around it. And I _think_ it's more performant. I want to try it out with comments and report back on that.

Comment (aaronlademann-wf) [reply test/over_react_component_test.html]: @jacehensley-wf IMO - if this script is necessary for the component to work when `props.useResizeObserver` is true, then the component should be responsible for dynamically loading / unloading that script into the DOM.

Comment (jacehensley-wf) [reply test/over_react_component_test.html]: But how does it know where the script is?

Comment (robbecker-wf) [reply test/over_react_component_test.html]: You could load it as a resource and inject it as a script?
See https://github.com/dart-lang/resource

Comment (jacehensley-wf) [reply test/over_react_component_test.html]: Interesting, I'll take a stab at that tonight

Comment (robbecker-wf) [reply test/over_react_component_test.html]: hmm I guess that makes more sense to do in the wrapper dart package

Comment (jacehensley-wf) [reply test/over_react_component_test.html]: The more I'm thinking about it the less I like the approach of injecting js without the user doing it. I think throwing an error and letting the user add the js file is the best approach.

**Pull request closed without merging**